### PR TITLE
[FSI] Two fields coupling and required refactoring

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**Description**
This PR adds support for two fields coupling (i.e. Interface Block Quasi-Newton equations). As this requires refactoring the non-linear iteration loop completely as well as other parts of the solver, I took the chance to completely refactor them to align them to the current design criteria.

Everything has been tested in serial / parallel (body fitted case) and embedded case. In a future PR I'll change the body fitted solver to use the ALE fluid one, which will simplify things further.

I create a draft PR as a bit more clean up is still required. However, this is in general terms ready for review.

**Changelog**
- Simplified embedded and body fitted FSI solvers
- Support for two fields coupling

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
Clean up finished. Ready to be reviewed.

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 1]
```diff
@@ -50,61 +96,155 @@ def GetInterfaceModelPart(self):
...
-            KratosMultiphysics.OLD_RELAXED_DISPLACEMENT,
+            self._old_relaxed_variable,
             iteration_value_vector)
 
-        # Compute the current non-linear iteration interface residual using the output variable
-        # Note that the residual is computed as r^{k+1} = \tilde{u}^{k+1} - u^{k}
-        output_variable_residual_vector = KratosMultiphysics.Vector(residual_size)
-        self._get_partitioned_fsi_utilities().ComputeInterfaceResidualVector(
+        # Compute the convergence accelerator correction
+        self._get_convergence_accelerator().UpdateSolution(self._output_variable_residual_vector, iteration_value_vector)
```
Comment (RiccardoRossi) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py:144]: is the indentation correct here??

File: applications/FSIApplication/python_scripts/partitioned_fsi_base_solver.py [hunk 2]
```diff
@@ -253,72 +211,75 @@ def SaveRestart(self):
...
             self._GetConvergenceAccelerator().InitializeNonLinearIteration()
+            if not self._GetConvergenceAccelerator().IsBlockNewton():
+                self._GetFSICouplingInterfaceStructure().Update()
+            else:
+                self._GetFSICouplingInterfaceStructure().UpdateDisplacement()
 
-            # Solve the mesh problem as well as the fluid problem
-            self._SolveMeshAndFluid()
+            # Update the structure interface position
+            self._GetFSICouplingInterfaceStructure().UpdatePosition(KratosMultiphysics.RELAXED_DISPLACEMENT)
```
Comment (RiccardoRossi) [inline applications/FSIApplication/python_scripts/partitioned_fsi_base_solver.py:239]: also here, is the indentation correct?

**Review by RiccardoRossi — COMMENTED** [review]
minor comments, please take a look, but otherwise accept

**Review by RiccardoRossi — DISMISSED** [review]
approving (conditional to taking a look to my two comments in the previous comment)

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp [hunk 3]
```diff
@@ -631,6 +616,64 @@ class PartitionedFSIUtilities
...
+        std::function<double(const double, const array_1d<double,3>)> traction_modulus_func;
+        if (SwapTractionSign)  {
+            traction_modulus_func = [](const double PosPressure, const array_1d<double,3>& rNormal){return - PosPressure / norm_2(rNormal);};
+        } else {
+            traction_modulus_func = [](const double PosPressure, const array_1d<double,3>& rNormal){return PosPressure / norm_2(rNormal);};
+        }
+
+        // Calculate the tractions from the pressure values
+        block_for_each(rModelPart.Nodes(), [&](Node<3>& rNode){
+            const auto& r_normal = rNode.FastGetSolutionStepValue(NORMAL);
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp]:
```suggestion
            const array_1d<double,3>& r_normal = rNode.FastGetSolutionStepValue(NORMAL);
```

remember we had issues with this

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp [hunk 4]
```diff
@@ -631,6 +616,64 @@ class PartitionedFSIUtilities
...
+        std::function<double(const double, const double, const array_1d<double,3>)> traction_modulus_func;
+        if (SwapTractionSign)  {
+            traction_modulus_func = [](const double PosPressure, const double NegPressure, const array_1d<double,3>& rNormal){return (NegPressure - PosPressure) / norm_2(rNormal);};
+        } else {
+            traction_modulus_func = [](const double PosPressure, const double NegPressure, const array_1d<double,3>& rNormal){return (PosPressure - NegPressure) / norm_2(rNormal);};
+        }
+
+        // Calculate the tractions from the pressure values
+        block_for_each(rModelPart.Nodes(), [&](Node<3>& rNode){
+            const auto& r_normal = rNode.FastGetSolutionStepValue(NORMAL);
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp]:
```suggestion
            const array_1d<double,3>& r_normal = rNode.FastGetSolutionStepValue(NORMAL);
```

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp [hunk 5]
```diff
@@ -666,6 +709,23 @@ class PartitionedFSIUtilities
...
+    std::string GetSkinConditionName()
+    {
+        std::string element_name;
+        if (TDim == 2) {
+            element_name = "LineCondition2D2N";
+        } else {
+            element_name = "SurfaceCondition3D3N";
+        }
+
+        return element_name;
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp]:
```suggestion
        if (TDim == 2) {
            return "LineCondition2D2N";
        } else {
            return "SurfaceCondition3D3N";
        }
```

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp [hunk 6]
```diff
@@ -31,6 +31,7 @@
 #include "includes/ublas_interface.h"
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp:31]: overall yyou could move the impl to cpp

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp [hunk 7]
```diff
@@ -53,6 +53,7 @@ void AddTrilinosConvergenceAcceleratorsToPython(pybind11::module &m)
     using TrilinosBaseConvergenceAcceleratorType = ConvergenceAccelerator<TrilinosSparseSpaceType, TrilinosLocalSpaceType>;
 
     py::class_<TrilinosBaseConvergenceAcceleratorType, typename TrilinosBaseConvergenceAcceleratorType::Pointer>(m, "TrilinosConvergenceAccelerator")
+        .def(py::init<>())
```
Comment (philbucher) [inline applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp:56]:
```suggestion
```

we wanted to avoid this

File: applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp [hunk 8]
```diff
@@ -339,7 +339,7 @@ namespace Testing {
             *p_interface_vector);
 
         const double tolerance = 1.0e-8;
-        std::vector<double> expected_values = {-1.0, -1.0, -1.0, -1.0};
+        std::vector<double> expected_values = {1.0,1.0,1.0,1.0};
```
Comment (philbucher) [inline applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp:342]:
```suggestion
        std::array<double, 4> expected_values = {1.0,1.0,1.0,1.0};
```

File: applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp [hunk 9]
```diff
@@ -339,7 +339,7 @@ namespace Testing {
...
         for (unsigned int i = 0; i < expected_values.size(); ++i) {
             KRATOS_CHECK_NEAR(expected_values[i], (*p_interface_vector)[i], tolerance);
```
Comment (philbucher) [inline applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp]: vector check macro

File: applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp [hunk 10]
```diff
@@ -365,7 +365,7 @@ namespace Testing {
             *p_interface_vector);
 
         const double tolerance = 1.0e-8;
-        std::vector<double> expected_values = {0.0,0.0,-1.0,-1.0,-2.0,-2.0,-3.0,-3.0};
+        std::vector<double> expected_values = {0.0,0.0,1.0,1.0,2.0,2.0,3.0,3.0};
         for (unsigned int i = 0; i < expected_values.size(); ++i) {
             KRATOS_CHECK_NEAR(expected_values[i], (*p_interface_vector)[i], tolerance);
         }
```
Comment (philbucher) [inline applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp]: same

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 11]
```diff
@@ -1,10 +1,13 @@
 from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
Comment (philbucher) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py]:
```suggestion
```

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 12]
```diff
@@ -1,10 +1,13 @@
...
 
 # Importing the Kratos Library
 import KratosMultiphysics
 
 # Import applications
 import KratosMultiphysics.FSIApplication as KratosFSI
+if KratosMultiphysics.ParallelEnvironment.GetDefaultDataCommunicator().IsDistributed():
+    import KratosMultiphysics.TrilinosApplication as KratosTrilinos
 
 class FSICouplingInterface():
```
Comment (philbucher) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py]:
```suggestion
class FSICouplingInterface:
```

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 13]
```diff
@@ -1,10 +1,13 @@
 from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+import math
 
 # Importing the Kratos Library
 import KratosMultiphysics
 
 # Import applications
 import KratosMultiphysics.FSIApplication as KratosFSI
+if KratosMultiphysics.ParallelEnvironment.GetDefaultDataCommunicator().IsDistributed():
+    import KratosMultiphysics.TrilinosApplication as KratosTrilinos
```
Comment (philbucher) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py:7]: I wouldn't do this here ...

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 14]
```diff
@@ -154,9 +340,30 @@ def _get_partitioned_fsi_utilities(self):
...
+            elif domain_size == 3:
+                if self.scalar_output:
+                    return KratosFSI.PartitionedFSIUtilitiesDouble3D()
+                else:
+                    return KratosFSI.PartitionedFSIUtilitiesArray3D()
+            else:
+                raise Exception("Domain size expected to be 2 or 3. Got " + str(self.domain_size))
         else:
-            raise Exception("Domain size expected to be 2 or 3. Got " + str(self.domain_size))
+            self._epetra_communicator = KratosTrilinos.CreateCommunicator()
```
Comment (philbucher) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py:356]: ... but here

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py [hunk 15]
```diff
@@ -154,9 +340,30 @@ def _get_partitioned_fsi_utilities(self):
...
+            if domain_size == 2:
+                if self.scalar_output:
+                    return KratosTrilinos.TrilinosPartitionedFSIUtilitiesDouble2D(self._epetra_communicator)
+                else:
+                    return KratosTrilinos.TrilinosPartitionedFSIUtilitiesArray2D(self._epetra_communicator)
+            elif domain_size == 3:
+                if self.scalar_output:
+                    return KratosTrilinos.TrilinosPartitionedFSIUtilitiesDouble3D(self._epetra_communicator)
+                else:
+                    return KratosTrilinos.TrilinosPartitionedFSIUtilitiesArray3D(self._epetra_communicator)
```
Comment (philbucher) [inline applications/FSIApplication/python_scripts/fsi_coupling_interface.py:366]: I think it is time to create a TrilinosExtension for the FSIApp
it is pretty straight forward (ofc not in this PR)

**Review by philbucher — COMMENTED** [review]
please revert the change in the TrilinosApp, other than that I am ok
nice cleanup :+1:

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/python_scripts/fsi_coupling_interface.py:144]: I think it is. For each call of the `Update` method we call the convergence accelerator `UpdateSolution` to correct the initial guess (note that if the indentation wouldn't be correct we'll get a runtime error in this case).

File: applications/FSIApplication/python_scripts/partitioned_fsi_base_solver.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/python_scripts/partitioned_fsi_base_solver.py:239]: I think it is. For each non-linear iteration you update the position of the coupling interface with the corrected solution (in this case stored in the `RELAXED_DISPLACEMENT` variable).

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp]: Yes! Thanks for spotting.

File: applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/partitioned_fsi_utilities.hpp:31]: Agree. We can do it in a standalone PR.

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py (hunk 15 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/python_scripts/fsi_coupling_interface.py:366]: Totally agree. Indeed I think is the only app missing. This is in top of my TODO lists now.

File: applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp (hunk 9 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp]: OK.

File: applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp (hunk 10 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/cpp_tests/test_partitioned_fsi_utilities.cpp]: OK.

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp (hunk 7 above)
Comment (philbucher) [reply applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp:56]: ping

Comment (rubenzorrilla) [reply applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp:56]: Recovering the discussion in https://github.com/KratosMultiphysics/Kratos/pull/8481#discussion_r594416796.

The problem is that the MPI convergence accelerators require the interface model part while the serial ones does not. As the model part in which the FSI problem minimization is to be performer is the one owning the convergence accelerator (the FSICouplingInterface) we need to first construct a fake convergence accelerator to then substitute it with the `SetConvergenceAccelerator` method.

We could of course create a fake convergence accelerator with empty parameters object, but in my opinion this would be rather hacky.

Long story short:
1. We create the fake base convergence accelerator with an empty parameters object.
2. We keep the default constructor exposure.

What do you think?

File: applications/FSIApplication/python_scripts/fsi_coupling_interface.py (hunk 14 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/python_scripts/fsi_coupling_interface.py:356]: Pf... As we have experienced in the past some issues with the MPI imports, I'd like to keep it in the top of the file.

Comment (philbucher) [reply applications/FSIApplication/python_scripts/fsi_coupling_interface.py:356]: I think those issues should be solved now but up to you

but be aware that as we discussed the GetDefault might be moved to the testing namespace so maybe it would be better to check the Kernel if it is a distributed run

**Review by philbucher — APPROVED** [review]
